editors/code/src/config.ts

```diff
@@ -97,96 +90,13 @@ export class Config {
-     * downloaded by the extension).
-     */
-    get prebuiltServerFileName(): null | string {
-        // See possible `arch` values here:
-        // https://nodejs.org/api/process.html#process_process_arch
-
-        switch (process.platform) {
-
-            case "linux": {
-                switch (process.arch) {
-                    case "arm":
-                    case "arm64": return null;
-
-                    default: return "rust-analyzer-linux";
-                }
-            }
-
-            case "darwin": return "rust-analyzer-mac";
-            case "win32": return "rust-analyzer-windows.exe";
-
-            // Users on these platforms yet need to manually build from sources
-            case "aix":
-            case "android":
-            case "freebsd":
-            case "openbsd":
-            case "sunos":
-            case "cygwin":
-            case "netbsd": return null;
-            // The list of platforms is exhaustive (see `NodeJS.Platform` type definition)
-        }
-    }
-
-    get installedExtensionUpdateChannel(): UpdatesChannel {
-        return this.extensionReleaseTag === NIGHTLY_TAG
-            ? UpdatesChannel.Nightly
-            : UpdatesChannel.Stable;
-    }
-
-    get serverSource(): null | ArtifactSource {
-        const serverPath = RA_LSP_DEBUG ?? this.serverPath;
-
-        if (serverPath) {
-            return {
-                type: ArtifactSource.Type.ExplicitPath,
-                path: Config.replaceTildeWithHomeDir(serverPath)
-            };
-        }
-
-        const prebuiltBinaryName = this.prebuiltServerFileName;
-
-        if (!prebuiltBinaryName) return null;
-
-        return this.createGithubReleaseSource(
-            prebuiltBinaryName,
-            this.extensionReleaseTag
-        );
-    }
-
-    private createGithubReleaseSource(file: string, tag: string): ArtifactSource.GithubRelease {
-        return {
-            type: ArtifactSource.Type.GithubRelease,
-            file,
-            tag,
-            dir: this.ctx.globalStoragePath,
-            repo: {
-                name: "rust-analyzer",
-                owner: "rust-analyzer",
-            }
-        };
-    }
-
-    get nightlyVsixSource(): ArtifactSource.GithubRelease {
-        return this.createGithubReleaseSource("rust-analyzer.vsix", NIGHTLY_TAG);
-    }
+    get globalStoragePath(): string { return this.ctx.globalStoragePath; }
 
     // We don't do runtime config validation here for simplicity. More on stackoverflow:
     // https://stackoverflow.com/questions/60135780/what-is-the-best-way-to-type-check-the-configuration-for-vscode-extension
 
-    private get serverPath() { return this.cfg.get("serverPath") as null | string; }
-    get updatesChannel() { return this.cfg.get("updates.channel") as UpdatesChannel; }
+    get serverPath() { return this.cfg.get("serverPath") as null | string; }
+    get channel(): "stable" | "nightly" { return this.cfg.get("updates.channel") ?? "stable"; }
```

Comment:
```suggestion
    get channel() { return this.cfg.get("updates.channel") as UpdatesChannel; }
```
The default `stable` should be returned due to the declaration in `package.json` configurations contribution point.
And I'd like to preserve the explicit `as` cast, since this is a downcast `any -> UpdatesChannel`

Reply:
>And I'd like to preserve the explicit as cast, since this is a downcast any -> UpdatesChannel

The type of get is

```TypeScript
get<T>(section: string): T | undefined;
```

So, this is explicitely to get rid of `|undefined` I guess an `!` would also work?

Reply:
I guess we should change all other configs to assert non undefined instead of `as`, but that's for other PR

Reply:
Let me elaborate here. The design of this `get<T>()` method is quite unsound. It lets you pass a type argument that will denote its return type.
```ts
const string | undefined = cfg.get<string>("name");
```
But that is very deceiving. By looking at this code you won't expect it to fail here since there is nothing unsound in it at the first glance.

However, what happens under the hood is the unchecked downcast `any -> T`.
Here in vscode impl you can see that this function returns a variable of type `any`:
![image](https://user-images.githubusercontent.com/36276403/76960373-fe555b00-6923-11ea-8b7e-60f55f448b3f.png)

And `any` is considered harmful, since it is implicitly convertible to literally any type (instead of `any` it is now recommended to use `unknown`).

So the user does can pass a value of some other type other than declared in our json schema in `package.json`  and this will fail at runtime, i.e.: see this comment
https://github.com/rust-analyzer/rust-analyzer/blob/2a3543d1953daeb240c214ececd273f1040b0516/editors/code/src/config.ts#L186-L187

My concern is to mark such unsound casts explicitly with `as` instead of silently passing the type arguments to the function that does an implicit cast under the hood so we don't even notice the unsoundness.

I know you are used to trusting `vscode` guys, but I *don't*. They are the same people as we and they do the same mistakes as we too and this is one of their bad design decisions because the proper signature for this function would be
```
get(opt: string): unknown;
```
or they should throw an error if the input doesn't match the declared json schema type in `package.json`.

Anyway, I'll at other the code later today!

Reply:
Note: type arguments, `!` null assertion and `as T` casts have no influence on the runtime behavior. They are just to silence the compiler complaints.

Reply:
This is out of scope of this PR, but I'd prefer to just run with upstream API. If they cast `any` to `T`, then they probably have the reasons to do so. In general, VS Code code base is of impecable code quality, so, in general, I trust them blindly when it comes to TypeScript. 

For the case of settings in the editor, I independently think that lenient validation is the right approach. It's more important that the thing works with broken settings, than to bail loudly on errors, *in this context*.